### PR TITLE
Implemented local editing of meal servings

### LINKED ISSUE #35: As a user, I want to be able to edit the number of servings of a meal on my meal plan from the home screen so that I can update my meal plan with changes to my lifestyle

### PR DESCRIPTION
**Pull request by adurso3**
Closes #35. Also fixes meal deletion.

### CODE REVIEW

File: app/src/main/java/com/mealsmadeeasy/ui/home/glance/WeekAtAGlanceFragment.kt
Comment (andrewbailey) [inline app/src/main/java/com/mealsmadeeasy/ui/home/glance/WeekAtAGlanceFragment.kt]: This is getting kinda complicated, so let's pull these two lambdas out into functions and refer to them using method references.

File: app/src/main/java/com/mealsmadeeasy/ui/home/glance/WeekAtAGlanceMealViewHolder.kt
Comment (andrewbailey) [inline app/src/main/java/com/mealsmadeeasy/ui/home/glance/WeekAtAGlanceMealViewHolder.kt]: Do you care about persisting this dialog across configuration changes? 'Cause this dialog will disappear if a user rotates their phone.

File: app/src/main/res/layout/view_edit_servings.xml
Comment (andrewbailey) [inline app/src/main/res/layout/view_edit_servings.xml]: This `RelativeLayout` is useless. Much like myself, it has no id, no style, and no background.

Unless you're trying to horizontally center its child, but if you don't have a container than Android doesn't inflate the LinearLayout's layout params correctly because you don't have a parent view for the dialog. In which case, use a `FrameLayout` here because it's faster.

File: app/src/main/res/values/strings.xml
Comment (andrewbailey) [inline app/src/main/res/values/strings.xml]: Maybe add ellipses 'cause you're opening a dialog? (If you add this, use `&#8230;` instead of `...`, because it looks prettier)

File: app/src/main/java/com/mealsmadeeasy/ui/home/glance/WeekAtAGlanceFragment.kt
Comment (andrewbailey) [inline app/src/main/java/com/mealsmadeeasy/ui/home/glance/WeekAtAGlanceFragment.kt]: wat. how did this happen. are you using this? how is this even resolved?

File: app/src/main/java/com/mealsmadeeasy/ui/home/glance/WeekAtAGlanceMealViewHolder.kt
Comment (andrewbailey) [inline app/src/main/java/com/mealsmadeeasy/ui/home/glance/WeekAtAGlanceMealViewHolder.kt]: Let's give this a title that includes the name of the meal we're adjusting the servings for. If you want to use a string resource, For better translations, feel free to make the string resource have a value of `Edit servings for "%1$s"`. `getString()` also takes in vararg objects that fill in these placeholders, much like `sprintf` in C.

File: app/src/main/java/com/mealsmadeeasy/ui/home/glance/WeekAtAGlanceMealViewHolder.kt
Comment (andrewbailey) [inline app/src/main/java/com/mealsmadeeasy/ui/home/glance/WeekAtAGlanceMealViewHolder.kt]: When changing the number of servings and returning to this dialog, it resets to the original value. Is there an issue here or with the meal store?

**Review by andrewbailey — CHANGES_REQUESTED** [review]
😢 It doesn't look like everything's working correctly. I can't tell if the serving size is being changed.

**Comment by adurso3** [comment]
I'll talk to you today / tomorrow about outstanding changes since I apparently don't extract methods into lambdas the way you want me to. I also don't seem to be having a problem with the mealstore servings displaying incorrectly. Screen rotation hasn't been done, but I wasn't sure we cared about that anymore. I agree that I should probably just go ahead and do it though.

**Comment by andrewbailey** [comment]
I don't remember what we decided here. I know we were going to address configuration changes, but did you want to do that in this PR or hold off until this gets merged?

File: app/src/main/java/com/mealsmadeeasy/ui/home/glance/WeekAtAGlanceMealViewHolder.kt
Comment (andrewbailey) [reply app/src/main/java/com/mealsmadeeasy/ui/home/glance/WeekAtAGlanceMealViewHolder.kt]: Just kidding. I can't reproduce this issue. 🤷‍♀️

File: app/src/main/res/values/strings.xml
Comment (andrewbailey) [inline app/src/main/res/values/strings.xml]: It looks like the quotes need to be escaped. I think you can just put `\`'s before them, or if you want to be fancy, you can use the curly double quotes. Then this would become `&#8220;%1$s&#8221;` because HTML codes.

**Comment by adurso3** [comment]
Since you want another round of change(s) on this PR, I'll just go ahead and do the configuration changes too, when I fix this last change.

File: app/src/main/java/com/mealsmadeeasy/ui/home/glance/WeekAtAGlanceMealViewHolder.kt
Comment (andrewbailey) [inline app/src/main/java/com/mealsmadeeasy/ui/home/glance/WeekAtAGlanceMealViewHolder.kt]: Consider using a `lateinit var`.

File: app/src/main/java/com/mealsmadeeasy/ui/home/glance/WeekAtAGlanceMealViewHolder.kt
Comment (andrewbailey) [inline app/src/main/java/com/mealsmadeeasy/ui/home/glance/WeekAtAGlanceMealViewHolder.kt]: You can make this super fancy and concise:
```kotlin
frag.arguments = Bundle().apply {
    putString(KEY_SAVED_DIALOG_TITLE, title)
    putInt(KEY_SAVED_SERVINGS, numServings)
}
```

File: app/src/main/java/com/mealsmadeeasy/ui/home/glance/WeekAtAGlanceMealViewHolder.kt
Comment (andrewbailey) [inline app/src/main/java/com/mealsmadeeasy/ui/home/glance/WeekAtAGlanceMealViewHolder.kt]: This name doesn't correspond to when it's used (`onAction` implies it's called whenever `Action` occurs)

File: app/src/main/java/com/mealsmadeeasy/ui/home/glance/WeekAtAGlanceMealViewHolder.kt
Comment (andrewbailey) [inline app/src/main/java/com/mealsmadeeasy/ui/home/glance/WeekAtAGlanceMealViewHolder.kt]: This needs to be implemented. You probably knew that already though...

**Review by andrewbailey — CHANGES_REQUESTED** [review]
Configuration changes are working, but editing functionality isn't working atm.

File: app/src/main/java/com/mealsmadeeasy/ui/home/glance/WeekAtAGlanceMealViewHolder.kt
Comment (adurso3) [reply app/src/main/java/com/mealsmadeeasy/ui/home/glance/WeekAtAGlanceMealViewHolder.kt]: I would but they won't let me. 😢

**Comment by adurso3** [comment]
Should be working now. ~~Please don't look at the code. It works, okay?~~

File: app/src/main/java/com/mealsmadeeasy/model/Meal.kt
Comment (andrewbailey) [inline app/src/main/java/com/mealsmadeeasy/model/Meal.kt]: Nit: `source` is non-nullable, so axe the `?` and the `!!` below.

Also 🎉 Parcelable! 🎉

File: app/src/main/java/com/mealsmadeeasy/model/Meal.kt
Comment (andrewbailey) [inline app/src/main/java/com/mealsmadeeasy/model/Meal.kt]: These methods can also probably use the single statement syntax: `override fun newArray(size: Int) = arrayOfNulls<Meal>(size)`

File: app/src/main/java/com/mealsmadeeasy/model/Meal.kt
Comment (andrewbailey) [inline app/src/main/java/com/mealsmadeeasy/model/Meal.kt]: Same about the one-line syntax here.

File: app/src/main/java/com/mealsmadeeasy/model/Meal.kt
Comment (andrewbailey) [inline app/src/main/java/com/mealsmadeeasy/model/Meal.kt]: dest is also non-nullable

File: app/src/main/java/com/mealsmadeeasy/model/MealPortion.kt
Comment (andrewbailey) [inline app/src/main/java/com/mealsmadeeasy/model/MealPortion.kt:9]: Same comments about the `Parcelable` implementation

File: app/src/main/java/com/mealsmadeeasy/ui/home/glance/EditServingsFragment.kt
Comment (andrewbailey) [inline app/src/main/java/com/mealsmadeeasy/ui/home/glance/EditServingsFragment.kt]: This doesn't look right. I think you want to axe the `args` variable and just do:
```
frag.arguments = Bundle().apply {
    putParcelable(KEY_SAVED_MEALPORTION, mealPortion)
    // others
}
```

File: app/src/main/java/com/mealsmadeeasy/ui/home/glance/EditServingsFragment.kt
Comment (andrewbailey) [inline app/src/main/java/com/mealsmadeeasy/ui/home/glance/EditServingsFragment.kt]: Opinions on using `!!`? I generally like to avoid it, but its behavior is appropriate in this context.

File: app/src/main/java/com/mealsmadeeasy/ui/home/glance/EditServingsFragment.kt
Comment (andrewbailey) [reply app/src/main/java/com/mealsmadeeasy/ui/home/glance/EditServingsFragment.kt]: (I think this is why the app is crashing)

**Review by andrewbailey — CHANGES_REQUESTED** [review]
Trying to edit servings crashes the app 😢

**Comment by andrewbailey** [comment]
I went ahead and pushed in some changes that fix the issues you ran into where editing serving sizes would result in duplicate meals and incredibly large serving sizes. There were two separate issues here:
1. When the dialog went to update serving counts, it called `removeMeal` and then `addMeal`. This caused a race condition because these methods are asynchronous, so `addMeal` wasn't guaranteed to start or end before `removeMeal` completed. This was fixed by adding an `editServingSize` method.
2. Time zones. Because of course time zones. It's always time zones. Basically, every time sent to the server is now sent in UTC. This should hopefully fix a few other outstanding issues.

Give a 👍 here when you look at these changes, and I'll go ahead and merge this.

**Comment by andrewbailey** [comment]
We saw an issue today where meals may still be duplicated on the server. I couldn't reproduce this by adding meals and changing their serving size. Is there a step I'm missing? Or did Studio perhaps not build and install the changes?

**Comment by adurso3** [comment]
Looks good. I give you my permission to merge your pull request into master, since I can't, lol. ~~Also, that Javadoc is 👌~~ I will now go reset the clock on my phone.

Edit: I'm very impressed that GitHub actually strikesthrough emoji.

**Review by andrewbailey — APPROVED** [review]

**Pull request merged**
